### DISCUSSION TITLE
3.6.0 Content

### DISCUSSION THREAD

**Discussion by bevans2000** [mars-sim/mars-sim#836] [General]
With the release of 3.5.0 imminent what should be the contents of release 3.6.0? We need the usual mixture of fixes/improvements/new features. I have started tagging some existing Issues with the 3.6.0 Milestones but we lack any significant new features.

**Comment by mokun** [comment]
Shall we begin to tackle on the `Project` feature, namely, revamping the Task/MetaTask and Mission to fall under it ?

**Reply by bevans2000** [reply]
Yes very much so.

**Comment by mokun** [comment]
Or should we start dabbling with the following MAJOR FEATURES : 

EDIT :
1. Convert to JSON serialization
2. Enable multi-floor layouts within a building
3. Model underground frozen water reserve
4. Build lava cave / crater cliff dwelling
5. Pave roads and remember existing routes
6. Allow zooming in maps in Mars Navigator 
7. Add higher resolution of Mars Map
8. Use html 5 compliant help browser

**Reply by bevans2000** [reply]
I like 6 as this would  be very useful and I assume we need 7 as well

**Comment by mokun** [comment]
Also, I can think of a few much needed MINOR IMPROVEMENTS : 

1. Make good use of the Settlement Map marquee banner
2. Add back notification msg boxes
3. Design a "collapsible" weather data panel in Settlement Map
4. Implement settlement time zone 
5. Schedule EVA tasks
6. Allow vehicles the option of docking with garage (thus no need of going inside)
7. Create aerial port/pad for drones to land and be fetched for loading/unloading
8. Add "All" in Settlement's Combo box in top center for certain tabs such as Citizens, Robots, Vehicle, Crop
9. Add back zoom ability in Orbit Viewer
10. Make use of CalendarProgram class (currently not being used) for choosing future resupply missions
11. Add ability for players to position an incoming building from a resupply mission

**Reply by bevans2000** [reply]
We have done some of 4 in the new Shift logic.
6 and 7 would be useful

**Reply by mokun** [reply]
We may also deepen the aspect of how robots will prepare the base in anticipation of human landing.

See this part of the [video](https://youtu.be/2tr6rJrv8hs?t=553) on what the scenario will visually look like.

Potentially, there's a lot of things we can fill in as to how to excavate regolith and how to use in-situ materials to build up a base.

@CiberNin laid some of those details out in that [thread](https://github.com/mars-sim/mars-sim/discussions/756#discussioncomment-4386102).

**Reply by mokun** [reply]
How do you like switching from html 4.2 to html 5 compliant in formmating help files ?

**Reply by bevans2000** [reply]
Yes, what benefit does it give to the end user?

**Comment by bevans2000** [comment]
For me, Construction is too hard. Why isn't more Construction done on the bigger Settlements?

**Reply by mokun** [reply]
> For me, Construction is too hard. Why isn't more Construction done on the bigger Settlements?

So let's find out that given enough raw resources if we can build a foundation, a frame and a building successfully.

Btw, do we have expert mode command on the console to add resources already ?

We should debug if the whole construction process work or not. This is one of the area being tested the least so far.

After we fix the bug, how about we put together a pop-up construction panel to explicitly show what   and how much resources are needed during a construction phase.

We should also show the reasons why a construction phase is pause and is waiting or what action to take place.

This panel will greatly aid players during a construction mission.

**Reply by bevans2000** [reply]
I will check. I know it is possible to add Resource to a Vehicle, I can't remember if it works at the ```ResourceHolder``` interface or just Vehicles.

**Reply by bevans2000** [reply]
To build a new Building it needs a  number of Construction Mission to build the various parts. So does that mean we need an overall Project to delivery a "new Greenhouse" for example? Maybe something related to #837 ?

**Reply by mokun** [reply]
Yay we could "simplify" construction of new building with this future "project" architecture.

**Comment by mokun** [comment]
By the way, we need a graphical interface for picking which position the player would like for each building to be shipped in via Resupply Tool.

In version v3.08 or v3.1.0, we tried creating what's called Transport Wizard to do that but it was very buggy. We gave up the idea. 

I recalled one difficult aspect was getting the event handling correctly, namely the keyboard and mouse control, in positioning the newly arrived building. 

We'll need to pause the simulation in order for the building to be added/inserted into the simulation. We also had to cancel all the walking tasks and recomputed them in light of the new building.

Would you like to try reimplementing the ability for player to position a building ?

**Reply by bevans2000** [reply]
Sounds like this should be captured as a new Issue.

**Comment by mokun** [comment]
So it turns out we mainly worked on maps in pre-3.6.0.

Also, you're wrapping up `Project` as well.

What do you see you must finish prior to the official release of 3.6.0 ?

**Reply by bevans2000** [reply]
I think we need the navigator window to be resizable as per #852. I haven't checked out your fix yet.

**Reply by mokun** [reply]
> I think we need the navigator window to be resizable as per #852. I haven't checked out your fix yet.

I treat that thread as making sure MissionWindow's nav tab panel contains no more ugly black padding space. 

One way to do it is simply making it having a fixed size.

Currently, the double int array in used is 450x450 in both Tools (NavigatorWindow and MissionWindow). 

If you stretch the original mapbox size of 450x450 into 600x300, then that array has to be redone in order to render a mapbox of that size.

It's still doable but the currently rendering with 5 map layers has been very slow on my 7th gen intel i7 8 cores laptop.

We may need to resolve how to concurrently load tiles before we make the mapbox stretchable and potentially making the size much bigger than our computer can handle well.

**Reply by bevans2000** [reply]
Also the Project work is background and I don't think it will be ready in the short term. But I did a good bit of work on it yesterday focussing on the Test drive mission and creating some common logic for Missions. dd3e897
It's feeling a lot simpler than the existing layered subclass approach.

**Comment by bevans2000** [comment]
@mokun we now have a lot of open issues that I think can be closed. Can we close some of them?

**Comment by bevans2000** [comment]
We should look to lock down what we want to complete as part of 3.6. There are new issues so let's add those we think should be included.

I want to complete #650 and the other related issues

**Reply by mokun** [reply]
> complete #650 and the other related issues

But #650 is going to be a big one, right ?

We do want to wrap up project mission, don't we ?

Anything still not being settled regarding the map resolution ?

**Reply by bevans2000** [reply]
Yes but #650 can be done gradually, not a big bang.

The Mission changes is similar, we don't have to do all at the same time

**Reply by mokun** [reply]
So MarsTime is done now. And it's also free of EVA suit bugs. What else for 3.6.0 ?

**Reply by mokun** [reply]
Let's close some in https://github.com/mars-sim/mars-sim/milestone/10 or and push some to the future.

**Reply by bevans2000** [reply]
I ha e near,y completed #650 and #888

**Comment by bevans2000** [comment]
I've closed done 3.6.0 issues but there are still a few open. We should look to close these down before starting anything new

**Reply by mokun** [reply]
The elevation resolution worths our attention. Do you want to have random local high resolution elevation data ?

**Reply by bevans2000** [reply]
Is this tagged for 3.6.0? I'm looking at #893 next. I think some of the open issues can be closed

**Reply by mokun** [reply]
> #893

On this issue, I've since added not just one dimension of relationship, but 3 aspects of relationship (namely, trust, care and respect) between any two individuals. 

However, we still need to take advantage of these 3 separate dimensions and need to distinguish when each level of relationship will be impacted.

**Reply by bevans2000** [reply]
Ok. I'll take it into account

**Comment by bevans2000** [comment]
@mokun I think that most of the remaining Open 3.6.0 issue have been done. Can we close them and then package 3.6.0 ?

**Comment by mokun** [comment]
Can you see if the compiled headless edition jar can be run currently ? I have not set up my machine to verify that.

**Reply by bevans2000** [reply]
Headless looks fine. I'm on holiday next week so won't be around.

**Comment by mokun** [comment]
Anyway else to close up ?

We can publish 3.6.0 today if the compiled swing and headless jars have no major issues.

**Reply by mokun** [reply]
3.6.0 has just been released. We can move onto 3.7.0 now.
